### PR TITLE
Parsing changes for Pre-Attribution Filtering: Attribution Scope

### PR DESCRIPTION
**Pull request by feifeiji89**
The current attribution logic in the Attribution Reporting API may not be ideal for use-cases where an ad-tech needs more fine grain control over the attribution granularity (i.e. campaign, product, conversion ID, etc.) before a source is chosen for attribution. Currently available features such as top-level filters are not fully sufficient for this use-case because they happen after a source has been selected (i.e. after attribution). We can optionally support this use-case by allowing callers of the API to specify predefined attribution scopes that will be considered for filtering before attributing a source, in order to more efficiently extract utility out of the API.

This PR covers the parsing logic for the new API surface.

***
<a href="https://pr-preview.s3.amazonaws.com/feifeiji89/attribution-reporting-api/pull/1352.html" title="Last updated on Jul 8, 2024, 4:14 PM UTC (3841d5f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/1352/bacfb84...feifeiji89:3841d5f.html" title="Last updated on Jul 8, 2024, 4:14 PM UTC (3841d5f)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -3165,6 +3229,10 @@ To <dfn>process an attribution source</dfn> given an [=attribution source=] |sou
 1. Set |source|'s [=attribution source/number of event-level reports=] to 0 if
     |source|'s [=attribution source/randomized response=] is null,
     [=attribution source/randomized response=]'s [=list/size=] otherwise.
+1. Let |states| be the [=obtain a set of possible trigger states|number of possible trigger states=] obtained with |randomizedResponseConfig|.
```
Comment (linnan-github) [inline index.bs:3232]: Can this be moved to the follow up PR as it’s not about parsing.

File: ts/src/header-validator/validate-json.ts [hunk 2]
```diff
@@ -1335,6 +1394,44 @@ function eventTriggerData(
...
+      return false
+    }
+    return true
+  }
+
+  return set(
+    j,
+    ctx,
+    (j) => string(j, ctx).filter(attributionScopeStringLength),
+    {
```
Comment (linnan-github) [inline ts/src/header-validator/validate-json.ts:1429]: minLength should be 1?

**Review by linnan-github — CHANGES_REQUESTED** [review]

File: ts/src/header-validator/source.test.ts [hunk 3]
```diff
@@ -2544,6 +2559,222 @@ const testCases: TestCase[] = [
...
+    name: 'attribution-scopes-too-many',
+    json: `{
+      "destination": "https://a.test",
+      "attribution_scope_limit": 1,
+      "attribution_scopes": ["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21"]
+    }`,
+    expectedErrors: [
+      {
+        path: ['attribution_scopes'],
+        msg: 'length must be in the range [0, 20]',
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply ts/src/header-validator/source.test.ts:2760]: It looks like the length is checked in two places and different error msg may be emitted depending on the error. That seems a little confusing, but probably fine? Cc @apasel422

[the remark this answers is not in the document]
Comment (feifeiji89) [reply ts/src/header-validator/source.test.ts:2760]: > Shouldn’t this be in the range [1, 1]?

The check against attribution_scope_limit is covered in test : `attribution-scopes-size-exceeds-attribution-scope-limit`

I updated the error message to be more accurate.

File: ts/src/header-validator/validate-json.ts [hunk 4]
```diff
@@ -1132,12 +1169,34 @@ function source(j: Json, ctx: SourceContext): Maybe<Source> {
           'destination_limit_priority',
           withDefault(int64, 0n)
         ),
+        attributionScopeLimit: field(
+          'attribution_scope_limit',
+          withDefault(positiveUint32, null)
+        ),
+        attributionScopes: field(
```
Comment (linnan-github) [inline ts/src/header-validator/validate-json.ts:1176]: Wdyt about passing the parsed `attributionScopeLimit` to validate `attribution_scopes` and `max_event_states`?

File: ts/src/header-validator/validate-json.ts [hunk 5]
```diff
@@ -759,6 +767,32 @@ function channelCapacity(s: Source, ctx: SourceContext): void {
...
+        return false
+      }
+      return true
+    }
+    return isInRange(
+      s.attributionScopes.size,
+      ctx,
+      1,
+      s.attributionScopeLimit,
+      `attribution_scope_limit must be greater than or equal to the number of attribution scopes ${s.attributionScopeLimit}`
```
Comment (apasel422) [inline ts/src/header-validator/validate-json.ts:791]:
```suggestion
      `attribution_scope_limit must be greater than or equal to the number of attribution scopes (${s.attributionScopeLimit})`
```

Comment (apasel422) [reply ts/src/header-validator/validate-json.ts:791]: But I'm also confused about the wording here, since it looks like we're validating the size of the `attribution_scopes` field, not the value of the `attribution_scope_limit` field. In other words, I'd expect this to be phrased as something like `number of attribution scopes (<number>) must be less than or equal to attribution_scope_limit (<number>)`.

File: index.bs [hunk 6]
```diff
@@ -2687,6 +2709,32 @@ non-negative integer |defaultBudget|, and an [=aggregatable debug reporting conf
...
+1. Let |values| be |map|["<code>[=source-registration JSON key/attribution_scopes=]</code>"].
+1. If |values| is not a [=list=], return an error.
+1. [=list/iterate|For each=] |value| of |values|:
+    1. If |value| is not a [=string=], return an error.
+    1. If |value|'s [=string/length=] is greater than [=max length of attribution scope for source=], return an error.
+    1. [=set/Append=] |value| to |result|.
+1. If |attributionScopeLimit| is null:
+    1. If |result| is not [=set/is empty|empty=], return an error.
+    1. Return |result|.
+1. If |result| [=set/is empty=], return an error.
```
Comment (apasel422) [inline index.bs:2733]: Could you clarify why it's an error for the scope set to be empty?

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: ts/src/header-validator/validate-json.ts (hunk 5 above)
Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts:791]: that's my initial thought as well. could you advise how do we change the field scope to attribution_scopes inside another scope? I was having trouble make it work somehow.

Comment (apasel422) [reply ts/src/header-validator/validate-json.ts:791]: You won't be able to *change* the scope within a `scope` call, only push a new one, so you'd have to check this outside the outer `ctx.scope` call.

File: index.bs (hunk 6 above)
Comment (feifeiji89) [reply index.bs:2733]: @linnan-github As Nan pointed in a separate thread, to prevent adtechs to select 2 buckets with scopes and no scopes when `attribution_scope_limit = 1`, which essentially gives adtech 2 scopes. so we've decided not to support empty scopes when the limit is specificied. cc'ed you the internal tracking bug.

File: ts/src/header-validator/validate-json.ts (hunk 5 above)
Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts]: thanks, as Nan suggested, I was able to remove this validation method after passing attributionScopeLimit value into `attributionScopes` and `maxEventStates`.

File: ts/src/header-validator/validate-json.ts (hunk 4 above)
Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts]: Done. thanks, it looks much cleaner

File: ts/src/header-validator/validate-json.ts [hunk 7]
```diff
@@ -1052,94 +1060,129 @@ export type Source = CommonDebug &
...
-            triggerSpecsFromTriggerData(j, ctx, triggerSpecsDeps),
-          ...(ctx.parseFullFlex
-            ? {
-                trigger_specs: (j) => triggerSpecs(j, ctx, triggerSpecsDeps),
-              }
-            : {}),
-        },
-        defaultTriggerSpecsVal
-      )(j, ctx)
+  return (
```
Comment (linnan-github) [inline ts/src/header-validator/validate-json.ts:1069]: Is outermost `()` necessary?

File: ts/src/header-validator/validate-json.ts [hunk 8]
```diff
@@ -1335,6 +1378,62 @@ function eventTriggerData(
...
+  return number(j, ctx)
+    .filter(isInteger, ctx)
+    .filter(isInRange, ctx, 1, ctx.vsv.maxTriggerStateCardinality)
+    .filter((n) => {
+      if (
+        attributionScopeLimit.value === null &&
+        n !== constants.defaultMaxEventStates
+      ) {
+        ctx.error(
+          'non-default max_event_states when attribution_scope_limit is not set'
```
Comment (linnan-github) [inline ts/src/header-validator/validate-json.ts:1401]: `max_event_states` in the message is probably redundant given it's the context.

File: index.bs (hunk 6 above)
Comment (apasel422) [reply index.bs:2733]: Please add a non-normative note explaining that below this algorithm, as it is not obvious.

File: ts/src/header-validator/validate-json.ts [hunk 9]
```diff
@@ -1052,94 +1060,129 @@ export type Source = CommonDebug &
...
+            'attribution_scopes',
+            withDefault(
+              (j) =>
+                attributionScopes(
+                  ctx,
+                  j,
+                  attributionScopeLimitVal.value !== null ? 1 : 0,
+                  Math.min(
+                    constants.maxAttributionScopesPerSource,
+                    attributionScopeLimitVal.value || 0
```
Comment (apasel422) [inline ts/src/header-validator/validate-json.ts:1163]:
```suggestion
                    attributionScopeLimitVal.value ?? 0
```

File: ts/src/header-validator/source.test.ts [hunk 10]
```diff
@@ -2544,6 +2559,223 @@ const testCases: TestCase[] = [
...
+  {
+    name: 'missing-attribution-scope-limit-max-event-states',
+    json: `{
+      "destination": "https://a.test",
+      "max_event_states": 5
+    }`,
+    expectedErrors: [
+      {
+        path: ['max_event_states'],
+        msg: 'non-default max_event_states when attribution_scope_limit is not set',
```
Comment (apasel422) [inline ts/src/header-validator/source.test.ts:2652]: Can we rephrase this to better indicate what the error actually is? It's unclear if the `when` means that this *is* the current situation, or that that *should be* the situation.

File: ts/src/header-validator/source.test.ts [hunk 11]
```diff
@@ -2544,6 +2559,223 @@ const testCases: TestCase[] = [
...
+  {
+    name: 'missing-attribution-scope-limit-attribution-scopes',
+    json: `{
+      "destination": "https://a.test",
+      "attribution_scopes": ["1", "2"]
+    }`,
+    expectedErrors: [
+      {
+        path: ['attribution_scopes'],
+        msg: 'length must be in the range [0, 0]',
```
Comment (apasel422) [inline ts/src/header-validator/source.test.ts:2639]: We may want to add an explanation to this message, as it isn't obvious why the list has to be empty. In particular, we may want to hint that the `attribution_scope_limit` field should be added to avoid this.

File: ts/src/header-validator/source.test.ts [hunk 12]
```diff
@@ -2544,6 +2559,223 @@ const testCases: TestCase[] = [
...
+    name: 'attribution-scopes-size-exceeds-attribution-scope-limit',
+    json: `{
+      "destination": "https://a.test",
+      "attribution_scope_limit": 1,
+      "attribution_scopes": ["1", "2"]
+    }`,
+    expectedErrors: [
+      {
+        path: ['attribution_scopes'],
+        msg: 'length must be in the range [1, 1]',
```
Comment (apasel422) [inline ts/src/header-validator/source.test.ts:2708]: We may want to indicate that the second `1` in this message is derived from the scope limit field.

File: ts/src/header-validator/source.test.ts [hunk 13]
```diff
@@ -2544,6 +2559,223 @@ const testCases: TestCase[] = [
...
+      },
+    ],
+  },
+  {
+    name: 'attribution-scope-not-string',
+    json: `{
+      "destination": "https://a.test",
+      "attribution_scope_limit": 2,
+      "attribution_scopes": [1, 2]
```
Comment (apasel422) [inline ts/src/header-validator/source.test.ts:2717]:
```suggestion
      "attribution_scopes": [1]
```
For simplicity.

File: ts/src/header-validator/trigger.test.ts [hunk 14]
```diff
@@ -1626,6 +1628,40 @@ const testCases: jsontest.TestCase<Trigger>[] = [
     ],
   },
 
+  // Attribution Scope
+  {
+    name: 'attribution-scope-not-string',
+    json: `{"attribution_scopes": [1, 2]}`,
```
Comment (apasel422) [inline ts/src/header-validator/trigger.test.ts:1634]:
```suggestion
    json: `{"attribution_scopes": [1]}`,
```
For simplicity.

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: ts/src/header-validator/validate-json.ts (hunk 7 above)
Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts:1069]: npm run pretty added this :)

File: index.bs [hunk 15]
```diff
@@ -2730,6 +2730,9 @@ To <dfn>parse attribution scopes for source</dfn> from a [=map=] |map| and a pos
 1. If |attributionScopeLimit| is null:
     1. If |result| is not [=set/is empty|empty=], return an error.
     1. Return |result|.
+
+Note: Empty attribution scopes are not allowed to prevent the selection of both buckets with and without scopes when |attributionScopeLimit| is set to 1, which would effectively result in two scopes.
```
Comment (apasel422) [inline index.bs:2734]: `bucket` seems to be a novel term here, and is easy to confuse with the same term from aggregatable attribution. Do we mean `sources` instead?

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: index.bs (hunk 6 above)
Comment (feifeiji89) [reply index.bs:2733]: Done.

File: ts/src/header-validator/source.test.ts (hunk 11 above)
Comment (feifeiji89) [reply ts/src/header-validator/source.test.ts]: refactored the code a bit to make error message more specific

File: index.bs [hunk 16]
```diff
@@ -2687,6 +2709,36 @@ non-negative integer |defaultBudget|, and an [=aggregatable debug reporting conf
...
+    1. If |result| is not [=set/is empty|empty=], return an error.
+    1. Return |result|.
+1. If |result| [=set/is empty=], return an error.
+1. If |result|'s [=set/size=] is greater than |attributionScopeLimit|, return an error.
+1. If |result|'s [=set/size=] is greater than [=max attribution scopes per source=], return an error.
+1. Return |result|.
+
+Note: Empty attribution scopes are not allowed if |attributionScopeLimit| is set,
+to prevent the selection of both buckets with and without scopes 
+if |attributionScopeLimit| is set to 1, which would effectively result in two scopes.
```
Comment (linnan-github) [inline index.bs:2740]: I don't think the issue only applies to `|attributionScopeLimit|` being 1. Instead, it applies to any value as that effectively increases the number of scopes by 1.

File: index.bs (hunk 15 above)
Comment (feifeiji89) [reply index.bs]: Good point. Done.

File: ts/src/header-validator/validate-json.ts [hunk 17]
```diff
@@ -1335,6 +1373,78 @@ function eventTriggerData(
...
+      )
+      return false
+    }
+    return true
+  }
+
+  return set(j, ctx, (j) =>
+    string(j, ctx).filter(attributionScopeStringLength)
+  ).filter((scopes) => {
+    if (attributionScopeLimit === undefined) {
```
Comment (linnan-github) [inline ts/src/header-validator/validate-json.ts:1422]: When would this happen?

Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts:1422]: at trigger parsing time.

File: index.bs (hunk 16 above)
Comment (feifeiji89) [reply index.bs]: lmk if it looks better now.

File: ts/src/header-validator/validate-json.ts [hunk 18]
```diff
@@ -1335,6 +1373,78 @@ function eventTriggerData(
...
+): Maybe<number> {
+  return number(j, ctx)
+    .filter(isInteger, ctx)
+    .filter(isInRange, ctx, 1, ctx.vsv.maxTriggerStateCardinality)
+    .filter((n) => {
+      if (
+        attributionScopeLimit.value === null &&
+        n !== constants.defaultMaxEventStates
+      ) {
+        ctx.error('non-default if attribution_scope_limit is not set')
```
Comment (linnan-github) [inline ts/src/header-validator/validate-json.ts:1395]: Maybe `non-default when attribution_scope_limit is not set`?

Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts]: good point. Done.

Comment (linnan-github) [reply ts/src/header-validator/validate-json.ts]: Please update the message to “must be default…” to avoid confusion.

Comment (feifeiji89) [reply ts/src/header-validator/validate-json.ts]: Done.

**Review by linnan-github — APPROVED** [review]

File: index.bs [hunk 19]
```diff
@@ -2687,6 +2709,36 @@ non-negative integer |defaultBudget|, and an [=aggregatable debug reporting conf
...
+1. If |attributionScopeLimit| is null:
+    1. If |result| is not [=set/is empty|empty=], return an error.
+    1. Return |result|.
+1. If |result| [=set/is empty=], return an error.
+1. If |result|'s [=set/size=] is greater than |attributionScopeLimit|, return an error.
+1. If |result|'s [=set/size=] is greater than [=max attribution scopes per source=], return an error.
+1. Return |result|.
+
+Note: Empty attribution scopes are not allowed if |attributionScopeLimit| is set,
+to prevent the selection of both sources with and without scopes which would effectively
```
Comment (apasel422) [inline index.bs:2739]:
```suggestion
to prevent the selection of both sources with and without scopes, which would effectively
```

**Review by apasel422 — APPROVED** [review]

**Review by apasel422 — APPROVED** [review]
